src/main/java/com/cloudbees/jenkins/plugins/mtslavescloud/MansionRetentionStrategy.java

```diff
@@ -72,32 +78,68 @@ private boolean shouldHaveConnectedByNow(T c) {
+        computer.setDisconnectInProgress(true);
+        // set accepting tasks so that others can see this flag if expecting it
         computer.setAcceptingTasks(false);
-        try {
-            Thread.sleep(2000);
-        } catch (InterruptedException e) {
-            LOGGER.log(Level.FINE, "Interrupted while sleeping before removing node " + n.getNodeName());
-            return;
-        }
-        if (!computer.isIdle() && computer.isOnline()) {
-            LOGGER.log(FINE, computer.getName() + " is no longer idle, aborting termination.");
-            // we lost the race -- mark it as back online
-            computer.setAcceptingTasks(true);
-            return;
-        }
-        for (Executor e : computer.getExecutors()) {
-            e.interrupt();
+        synchronized (this) {
+            if (disconnectInProgress) {
+                // if we already have a background thread running, then we can return immediately
+                return;
+            }
+            disconnectInProgress = true;
         }
-        LOGGER.log(Level.FINE, "Finally removing node " + n.getNodeName());
-        super.kill(n);
+        Computer.threadPoolForRemoting.submit(new Runnable() {
+            public void run() {
+                // TODO once Jenkins 1.607+ we can remove the sleep as the withLock will give atomic guarantee
+
+                // attempt to heuristically detect a race condition
+                // where an executor accepted a task after we checked for
+                // idleness,
+                // but before we marked it as unavailable for tasks
+                // See JENKINS-23676
+                try {
+                    Thread.sleep(2000);
+                } catch (InterruptedException e) {
+                    LOGGER.log(Level.FINE, "Interrupted while sleeping before removing node {0}", nodeName);
+                    return;
+                }
+                
+                queueLock.withLock(new Runnable() {
+                    public void run() {
+                        if (!computer.isIdle() && computer.isOnline()) {
+                            LOGGER.log(FINE, computer.getName() + " is no longer idle, aborting termination.");
+                            // we lost the race -- mark it as back online
+                            computer.setAcceptingTasks(true);
+                            computer.setDisconnectInProgress(false);
+                            synchronized (MansionRetentionStrategy.this) {
+                                disconnectInProgress = false;
+                            }
+                            return;
+                        }
+                        // TODO figure out why this cannot just be computer.getNode().terminate()
+                        try {
+                            for (Executor e : computer.getExecutors()) {
+                                e.interrupt();
```

Comment:
Well, I've seen cases where a node is dead, but executors are still "alive" and so couldn't aren't removed properly.